### PR TITLE
util: Add Assume() identity function

### PR DESCRIPTION
**Pull request by maflcko**
This is needed for #20138. Please refer to the added documentation for motivation.

### CODE REVIEW

**Comment by laanwj** [comment]
Concept ACK. It's mildly interesting that we're sort of reinventing the original `assert()` here, which gets disabled on `NDEBUG`, which we had to always force on, but I think it's fine to reinvent the wheel here instead of assuming any specific compiler semantics.

**Comment by MarcoFalke** [comment]
Indeed, this `Assume` is like the traditional `assert`.

**Comment by jnewbery** [comment]
Concept ACK. If you're willing to entertain paint color requests, I prefer the name `Check()`, as was originally proposed in #16136.

**Comment by MarcoFalke** [comment]
I think @ryanofsky also likes `check`, so I might change to that

**Comment by jnewbery** [comment]
I think Russ doesn't care about the name, but would prefer a pair of check/dcheck macros: https://github.com/bitcoin/bitcoin/pull/16136#issuecomment-545993867

**Comment by sipa** [comment]
FWIW, libsecp256k1 has CHECK() (always enabled) and VERIFY_CHECK() functions (enabled with -DVERIFY, used in test builds).

**Comment by practicalswift** [comment]
Concept ACK

**Comment by MarcoFalke** [comment]
> FWIW, libsecp256k1 has CHECK() (always enabled) and VERIFY_CHECK() functions (enabled with -DVERIFY, used in test builds).

That is good to know, because that means we can't use that name due to name-clashing.

**Review by ajtowns — COMMENTED** [review]
ACK fa9672603560320bc9f1017a7de5e85de10d9561

I think this is a good choice of behaviour that should be easy to use correctly. Would like to see the ability to turn on logging of violated Assumptions even in production builds in a followup.

**Comment by practicalswift** [comment]
ACK fa9672603560320bc9f1017a7de5e85de10d9561: patch looks good

**Review by hebasto — COMMENTED** [review]
Approach ACK fa9672603560320bc9f1017a7de5e85de10d9561

Are there any cases when `assert` is preferable to `Assert`. If "yes", it would be nice to have an example in the docs. If "no", why do we mention `assert` in the docs?

File: doc/developer-notes.md
Comment (MarcoFalke) [reply doc/developer-notes.md]: thx, done

File: doc/developer-notes.md
Comment (MarcoFalke) [reply doc/developer-notes.md:304]: > Maybe rephrase it as a directive

thx, done

> maybe we should deprecate assert()

I'll leave that for a follow-up

**Comment by MarcoFalke** [comment]
> Are there any cases when assert is preferable to Assert

No

> If "no", why do we mention assert in the docs?

It documents how the `assert` is used in existing code

**Comment by practicalswift** [comment]
cr ACK faa05854f832405231c9198787a4eafe2cd4c5f0

**Comment by jnewbery** [comment]
utACK faa05854f832405231c9198787a4eafe2cd4c5f0

File: doc/developer-notes.md
Comment (jonatack) [inline doc/developer-notes.md:294]: s/can not/cannot/

File: doc/developer-notes.md
Comment (jonatack) [inline doc/developer-notes.md:301]: s/For example it can/For example, it may/

**Review by jonatack — COMMENTED** [review]
Approach ACK, some doc nits if you retouch

**Comment by MarcoFalke** [comment]
Will leave style nits alone to preserve ACKs

**Comment by MarcoFalke** [comment]
I think this is ready, but I am waiting on one more (re?)ACK before merge

**Review by hebasto — APPROVED** [review]
ACK faa05854f832405231c9198787a4eafe2cd4c5f0, I have reviewed the code and it looks OK, I agree it can be merged.

**Pull request merged**
